### ISSUE TITLE
Class constants not supported

### ISSUE THREAD

**Issue by POPSuL** [ApiGen/ApiGen#877]
Example file:
```
class A
{
    const B = '123';
}
```

Output:
```
 % ~/projects/ApiGen/bin/apigen generate -vvv --destination=/var/www/api /var/www/test

  [ReflectionException (-1)]
  Class A does not exist


Exception trace:
 () at /home/popsul/projects/ApiGen/packages/Reflection/src/Reflection/Class_/ClassConstantReflection.php:46
 ReflectionClass->__construct() at /home/popsul/projects/ApiGen/packages/Reflection/src/Reflection/Class_/ClassConstantReflection.php:46
 ApiGen\Reflection\Reflection\Class_\ClassConstantReflection->__construct() at /home/popsul/projects/ApiGen/packages/Reflection/src/Reflection/Class_/ClassConstantReflection.php:58
 ApiGen\Reflection\Reflection\Class_\ClassConstantReflection::createFromNameValueAndClass() at /home/popsul/projects/ApiGen/packages/Reflection/src/Reflection/Class_/ClassReflection.php:226
 ApiGen\Reflection\Reflection\Class_\ClassReflection->getOwnConstants() at /home/popsul/projects/ApiGen/packages/Element/src/Cache/ReflectionWarmUpper.php:33
 ApiGen\Element\Cache\ReflectionWarmUpper->warmUpClassReflections() at /home/popsul/projects/ApiGen/packages/Element/src/Cache/ReflectionWarmUpper.php:24
 ApiGen\Element\Cache\ReflectionWarmUpper->warmUp() at /home/popsul/projects/ApiGen/packages/Reflection/src/Parser/Parser.php:63
 ApiGen\Reflection\Parser\Parser->parseDirectories() at /home/popsul/projects/ApiGen/src/Application/ApiGenApplication.php:55
 ApiGen\Application\ApiGenApplication->runCommand() at /home/popsul/projects/ApiGen/src/Console/Command/GenerateCommand.php:61
 ApiGen\Console\Command\GenerateCommand->execute() at /home/popsul/projects/ApiGen/vendor/symfony/console/Command/Command.php:264
 Symfony\Component\Console\Command\Command->run() at /home/popsul/projects/ApiGen/vendor/symfony/console/Application.php:869
 Symfony\Component\Console\Application->doRunCommand() at /home/popsul/projects/ApiGen/vendor/symfony/console/Application.php:223
 Symfony\Component\Console\Application->doRun() at /home/popsul/projects/ApiGen/vendor/symfony/console/Application.php:130
 Symfony\Component\Console\Application->run() at /home/popsul/projects/ApiGen/bin/apigen:17

generate [--destination DESTINATION] [--] <source> (<source>)...
```

**Comment by TomasVotruba** [comment]
How is class autoloaded?

**Comment by POPSuL** [comment]
@TomasVotruba in no way. I have directory with single file named A.php which contains class A with one constant.
This problem is caused by the use of native php reflection, which knows nothing about the source code.

**Comment by TomasVotruba** [comment]
Well, I think BetterReflection we use now depends on autoloading, own or composer.

Could you autoload this class by composer and re-run please? Let me know the result.

**Comment by POPSuL** [comment]
@TomasVotruba yes, if I preload file - all works properly.

**Comment by TomasVotruba** [comment]
At the moment, our [Parser used directories and vendor source locators](https://github.com/ApiGen/ApiGen/blob/b547c4ab426984f28041d622778a6dbed9f6ed98/packages/Reflection/src/Parser/Parser.php#L121-L132).

You can check all available locators, if there is something that would solve this:
https://github.com/Roave/BetterReflection/tree/master/src/SourceLocator/Type

Or ask issues at https://github.com/Roave/BetterReflection/, they've been very helpful to me.

**Comment by POPSuL** [comment]
@TomasVotruba I write example script for this reflection library:
```
require "vendor/autoload.php";

use Roave\BetterReflection\Reflector\ClassReflector;
use Roave\BetterReflection\SourceLocator\Type\DirectoriesSourceLocator;

$reflector = new ClassReflector(new DirectoriesSourceLocator(['/var/www/test']));
$reflectionClass = $reflector->reflect(A::class);

echo $reflectionClass->getName(), PHP_EOL;
var_dump($reflectionClass->getConstants());

var_dump(class_exists(A::class));
```

Output:
```
A
array(1) {
  ["B"]=>
  string(3) "123"
}
bool(false)
```

This library processed the file without autoloading (class_exists returns false).

**Comment by TomasVotruba** [comment]
@POPSuL Thanks for making such a nice example.

Could you try calling [`getAllClasses()`]
(https://github.com/ApiGen/ApiGen/blob/b547c4ab426984f28041d622778a6dbed9f6ed98/packages/Reflection/src/Parser/Parser.php#L102) as in ApiGen?

**Comment by POPSuL** [comment]
@TomasVotruba if I add `$reflector->getAllClasses()` to my example I get array with one object of type `Roave\BetterReflection\Reflection\ReflectionClass` which related to my `A::class`.

Imho, for solve this issue need remove usage of native reflection in https://github.com/ApiGen/ApiGen/blob/master/packages/Reflection/src/Reflection/Class_/ClassConstantReflection.php#L44
and rewrite `ClassReflection::getOwnConstants()` (https://github.com/ApiGen/ApiGen/blob/master/packages/Reflection/src/Reflection/Class_/ClassReflection.php#L225).
Yep, I know `ReflectionClass::getConstants()` from BetterReflection returns hashtable `string[string]`, but it can be rewritten. Need add some type (e.g. ClassConstantReflection) and transform ConstNode (which provide info about constant visibility) to that type and do `::getConstants(): ClassConstantReflection[string]`.

**Comment by TomasVotruba** [comment]
@POPSuL There is a solution then, that's great.

I got lost in links and method names. I feel like you have it thought trought. 
Could you please send PR with failing tests and fix?

**Comment by POPSuL** [comment]
Proposed PR related to this issue: https://github.com/Roave/BetterReflection/pull/282

**Comment by TomasVotruba** [comment]
Thank you :bowing_man:

**Comment by frontalnugity** [comment]
I'm still running into this issue - not sure what to do in order to be able to move forward other than commenting out the constants in my class object.  .I'm trying to override some base properties in my laravel root class to match my config.

Any suggestions or an eta on when this may be resolved or a proper workaround?  I see something in the thread above about 'preloading' but don't understand what steps are taken to do this.  Help! :)

```
<?php
namespace XMLT\Ignition\Base;
use Illuminate\Database\Eloquent\Model;
class BaseModel extends Model
{
	/**
      * The name of the "created at" column.
      *
      * @var string
      */
     const CREATED_AT = 'created';
	 /**
       * The name of the "updated at" column.
       *
       * @var string
       */
	 const UPDATED_AT = 'updated';
}
```

**Comment by TomasVotruba** [comment]
@frontalnugity I think the best way to go now is to support it here: https://github.com/Roave/BetterReflection/pull/282

Investing energy to workaround instead of solution that is in very nice state would slow down the propper solution.

**Comment by TomasVotruba** [comment]
@frontalnugity I think the best way to go now is to support it here: https://github.com/Roave/BetterReflection/pull/282

Investing energy to workaround instead of solution that is in very nice state would slow down the propper solution.

**Comment by TomasVotruba** [comment]
Beeing resolved in #926
